platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt

```diff
@@ -109,58 +110,77 @@ data class StructuredMessage(val fragments: List<Fragment>) {
 
 sealed class PropertyTrace {
 
-    object Unknown : PropertyTrace()
+    object Unknown : PropertyTrace() {
+        override fun toString(): String = asString()
+    }
 
-    object Gradle : PropertyTrace()
+    object Gradle : PropertyTrace() {
+        override fun toString(): String = asString()
+    }
 
-    class BuildLogic(
+    data class BuildLogic(
         val source: DisplayName,
```

Comment:
❔  My only concern here is expecting this very general interface as a field. Currently, it seems not to be a problem, but it might be possible in the future to accidentally pass something here that would not necessarily be designed for being a "data" object. I was looking at the `DisplayName` implementations and their abundance.

Reply:
Would providing factory methods or additional constructors that accepted `Location` and `UserCodeSource` (and then extracted display name and line number, then narrowing the kind of display names we accept here) alleviate your concern?

Reply:
I would say yes 👍